### PR TITLE
Added extraSprites prop to MapView (Reworked how images are passed to RNMBGL)

### PR DESCRIPTION
**Pull request by brandonreavis**
## What it Does 
This PR improves how images are registered with react-native-mapbox-gl by providing a new `extraSprites` prop on the `MapView` component and deprecating including image URLs (and ids) directly in the styles.

## The Problems it Solves 

The current allowed method of registering images with react-native-mapbox-gl is something like:

```js
import airportIcon from './icons/airportIcon.png';

const iconStyle = {
	iconImage: airportIcon
};

// render method
class MyMap extends React.Component {
	render() {
		return <MapboxGL.MapView styleURL="">
			<MapboxGL.VectorSource>
				<MapboxGL.SymbolLayer style={iconStyle}/>
			</MapboxGL.VectorSource>
		</MapboxGL.MapView>;
	}
}
```

This has a number of issues:
1. Using an image from the style's spritesheet by name is not allowed, because RNMBGL thinks strings and numbers are all URIs that need to be fetched.
2. If you want to use an expression for a prop like `iconImage` it can only reference images in the style's spritesheet because RNMBGL only loads images if `iconImage` is set to a number or string. Traversing an expression for URLs and loading them is a bad idea.
3. If you want to let the data decide what custom icon (that is not in the styles original spritesheet) to use (eg. `iconImage: ['get', 'icon']`), you can't, because RNMBGL requires the image URLs to be in the style.)

## Example of How it Works
This PR fixes all of that by decoupling image loading from the styles by adding an `extraSprites` property to the MapView.

```js
import airportIcon from './icons/airportIcon.png';

const extraSprites = {
	airport: airportIcon
}; 

const iconStyle = {
	iconImage: 'airport' // or ['get', 'icon'], or ['match', ['get', 'class'], 'aeroway', 'airport', ''], ...
};

class MyMap extends React.Component {
	render() {
		return <MapboxGL.MapView styleURL="" extraSprites={extraSprites}>
			<MapboxGL.VectorSource>
				<MapboxGL.SymbolLayer style={iconStyle}/>
			</MapboxGL.VectorSource>
		</MapboxGL.MapView>;
	}
}
```

This is a breaking change, but I think it is important enough that it would be great if it could make it into 7.0.0. Going forward with image handling for the library, I think it is best to go with something like this which is simple, less brittle, and closer to how the native SDK supports custom images.

### CODE REVIEW

**Comment by mfazekas** [comment]
@brandonreavis i think our existing [Images](https://github.com/react-native-mapbox-gl/maps/blob/master/docs/Images.md) and the old [ShapeSource#images](https://github.com/nitaliano/react-native-mapbox-gl/blob/master/docs/ShapeSource.md#images) serves the same purpose.

See the ShapeSourceIcon example:
https://github.com/react-native-mapbox-gl/maps/blob/939ccfc6a7d360f2679e84873d8dda9aa6b86ce5/example/src/examples/ShapeSourceIcon.js#L10-L24

https://github.com/react-native-mapbox-gl/maps/blob/939ccfc6a7d360f2679e84873d8dda9aa6b86ce5/example/src/examples/ShapeSourceIcon.js#L72-L88

**Comment by brandonreavis** [comment]
This PR addresses two main issues I have with how image loading in react-native-mapbox-gl. The new `<Images/>` component only touches on the first issue below.

**Issue 1: Style and Code Clarity**

`<MapView><Images images={}/></MapView>` vs `<MapView extraSprites={}/>`

Both approaches work, but having images be a prop of the MapView makes more sense to me for the following reasons:

1.  The `<Images/>` component just creates a dummy ShapeSource with a random id under the hood. Why image loading happens in the ShapeSource and not the MapView does not make sense to me. Natively, the MapView owns the images – not any of the sources. There are many cases for using images that have no need of a ShapeSource. Creating a dummy shape source to load images works, but definitely feels awkward.
2. Why is `Images` a child component and not just a property of the MapView? I know this a matter of personal preference, but using a child component seems weird for representing a setting on the MapView that doesn't have any inherent hierarchy. How is it any better than having a prop on the MapView itself?

That said, if it is being considered for merging, I need to get the PR to support bundled assets like ``{uri: 'asset:/app_icon.png'}``. It should accept anything `<Image source=...>` supports without any special syntax like the `{assets: ['pin']}`.

If you want to rename `extraSprites` to something like `images` I'm fine with that. 

**Issue 2: React Native Mapbox GL's Style JSON limitations**

TLDR: We really need to deprecate the ability to put image URIs in style JSON. 

This is more important to solve before another major release goes out and is less subjective than the previous issue. Your example didn't address these:

>>>1. Using an image from the style's spritesheet by name is not allowed, because RNMBGL thinks strings and numbers are all URIs that need to be fetched.
>>>2. If you want to use an expression for a prop like iconImage it can only reference images in the style's spritesheet because RNMBGL only loads images if iconImage is set to a number or string. 3. Traversing an expression for URLs and loading them is a bad idea.
>>>3. If you want to let the data decide what custom icon (that is not in the styles original spritesheet) to use (eg. iconImage: ['get', 'icon']), you can't, because RNMBGL requires the image URLs to be in the style.)

I think it is a good goal to have style JSON in mapbox gl be as close to 1-to-1 with the style spec from Mapbox (https://docs.mapbox.com/mapbox-gl-js/style-spec/#layout-symbol-icon-image). Right now if you do something like `iconImage: 'airport'` react-native-mapbox-gl tries to load an image at the path `'airport'` and fails. There have been recently been a series of commits to check for and ignore tokenized values like `{icon}`. There is no sane workaround for determining if a string literal is referring to a uri or a named image from the spritesheet. Whatever syntax wins for issue 1 above should be the one-and-only to use custom images for your map.

**Comment by mfazekas** [comment]
@brandonreavis `Image` is `ShapeSource` is really an implementation detail, the plan is to extract it to separate native component over the time. It's a shape source as before it was a property of shapesource layer.

Having a component is an advantage vs to property where you want to create a self contained component that might want to specify it's image sources. Say the`<UserLocation` component wants to use images for represeting the current user location, now it can, without requiring you to add something the `<MapView `-s `extraSprite` property.

I do like the idea of making `iconImage: 'foo'` use image named 'foo' vs trying to load 'foo' directly.
If someone does 'iconImage: require('foo.png')` that's something we can recognize, we can add special syntax if someone wants to load a path/url.

> * Using an image from the style's spritesheet by name is not allowed, because RNMBGL thinks strings and numbers are all URIs that need to be fetched.
> 
That is a valid issue, the workaround for now is `iconImage: ['literal', 'airport']` We should probably change this and `iconImage: 'airport'` should try to get the image named airport, vs trying to load.
`require` and `urls` could be still supported.

> * If you want to use an expression for a prop like iconImage it can only reference images in the style's spritesheet because RNMBGL only loads images if iconImage is set to a number or string. 3. Traversing an expression for URLs and loading them is a bad idea.

Not sure i get it. Expressions should resolve to an image name i, we don't try to load url-s from expressions.

> * If you want to let the data decide what custom icon (that is not in the styles original spritesheet) to use (eg. iconImage: ['get', 'icon']), you can't, because RNMBGL requires the image URLs to be in the style.)

You can use the Images component to specify image url-s.

**Comment by brandonreavis** [comment]
Well if we at least agree that any string or array value of an image prop of a style should be sent directly to Mapbox GL native without any special treatment, that is good. 

The breaking changes that that will introduce feels like a good opportunity to consolidate and simplify the logic so that all the image registration happens via one paradigm (ie. Images component), but if you feel like maintaining the image registration all over the place (all the style prop setters), more power to you. 

We can close this and that can be addressed separately. Thanks for maintaining the project.

**Comment by brandonreavis** [comment]
Regarding 
> we can add special syntax if someone wants to load a path/url.

I vote the special syntax matches what the RN Image component expects (ie. `{uri: '/path/to/image.png'}`)
